File: tests/metrics/total_return_test.py
from __future__ import annotations
import empyrical
import numpy as np
import pytest
from taflow.metrics.total_return import TotalReturn

@pytest.mark.parametrize('returns', [np.array([0.01]), np.array([0.1, -0.2, 0.05]), np.zeros(32), np.linspace(-0.04, 0.05, 101), np.array([np.nan, 0.02, -0.03, np.nan, 0.01]), np.random.default_rng(20260811).normal(0.0004, 0.012, 513)])
def test_total_return_matches_empyrical(returns: np.ndarray) -> None:
    actual = TotalReturn().from_returns(returns).compute()
    expected = float(empyrical.cum_returns_final(returns))
    assert actual == pytest.approx(expected, rel=1e-12, abs=1e-14)

def test_total_return_input_methods_and_lifecycle_are_invariant() -> None:
    returns = np.array([0.1, -0.2, 0.05])
    expected = TotalReturn().from_returns(returns).compute()
    equity = np.array([100.0, 110.0, 88.0, 92.4])
    pnl = np.array([10.0, -22.0, 4.4])
    assert TotalReturn().from_equity(equity).compute() == pytest.approx(expected)
    assert TotalReturn().from_pnl(pnl, initial_capital=100.0).compute() == pytest.approx(expected)
    assert TotalReturn().from_log_returns(np.log1p(returns)).compute() == pytest.approx(expected)
    state = TotalReturn().from_returns([])
    assert state.value is None
    assert state.append(returns[0]) is state
    assert state.extend(returns[1:]) is state
    assert state.compute() == pytest.approx(expected)
    assert len(state) == len(returns)
    assert state.reset() is state
    assert len(state) == 0
    assert state.extend(returns).compute() == pytest.approx(expected)

def test_total_return_missing_and_invalid_values() -> None:
    assert len(TotalReturn().from_returns([np.nan, 0.01])) == 1
    with pytest.raises(ValueError):
        TotalReturn(nan_policy='raise').from_returns([np.nan])
    with pytest.raises(ValueError):
        TotalReturn().from_returns([np.inf])
    with pytest.raises(ValueError):
        TotalReturn().from_returns([-1.01])

def test_total_return_requires_semantic_ingestion_before_append() -> None:
    with pytest.raises(ValueError):
        TotalReturn().append(0.01)

File: crates/taflow-metrics/src/metrics/total_return_test.rs
use super::total_return::TotalReturn;
use crate::{MetricInputKind, NanPolicy};

fn assert_close(actual: f64, expected: f64) {
    assert!((actual - expected).abs() <= 1e-14, "{actual} != {expected}");
}

#[test]
fn compounds_returns_and_preserves_lifecycle() {
    let values = [0.10, -0.20, 0.05];
    let expected = 1.10 * 0.80 * 1.05 - 1.0;
    let mut state = TotalReturn::new(NanPolicy::Omit).unwrap();
    state.from_returns(&[]).unwrap();

    assert_eq!(state.value(), None);
    state.extend(&values[..2]).unwrap();
    assert_close(state.append(values[2]).unwrap().unwrap(), expected);
    assert_close(state.compute().unwrap(), expected);
    assert_eq!(state.len(), 3);

    state.reset();
    assert!(state.is_empty());
    assert_eq!(state.value(), None);
    assert_close(state.extend(&values).unwrap().unwrap(), expected);
}

#[test]
fn input_modes_produce_equivalent_returns() {
    let returns = [0.10, -0.20, 0.05];
    let expected = 1.10 * 0.80 * 1.05 - 1.0;

    let mut equity = TotalReturn::new(NanPolicy::Omit).unwrap();
    equity.from_equity(&[]).unwrap();
    assert_close(
        equity.extend(&[100.0, 110.0, 88.0, 92.4]).unwrap().unwrap(),
        expected,
    );
    assert_eq!(equity.len(), 3);

    let mut pnl = TotalReturn::new(NanPolicy::Omit).unwrap();
    pnl.from_pnl(&[], 100.0).unwrap();
    assert_close(pnl.extend(&[10.0, -22.0, 4.4]).unwrap().unwrap(), expected);

    let log_returns = returns.map(f64::ln_1p);
    let mut logarithmic = TotalReturn::new(NanPolicy::Omit).unwrap();
    logarithmic.from_log_returns(&[]).unwrap();
    assert_close(logarithmic.extend(&log_returns).unwrap().unwrap(), expected);
}

#[test]
fn omits_nan_and_handles_total_loss() {
    let mut state = TotalReturn::new(NanPolicy::Omit).unwrap();
    state.from_returns(&[]).unwrap();
    state.extend(&[f64::NAN, 0.25, -1.0]).unwrap();
    assert_eq!(state.value(), Some(-1.0));
    assert_eq!(state.len(), 2);
}

#[test]
fn rejects_non_return_semantic_domains() {
    let mut unbound = TotalReturn::new(NanPolicy::Omit).unwrap();
    assert!(unbound.append(0.01).is_err());
}
